pkg/cmd/roachtest/run.go

```diff
@@ -327,3 +333,102 @@ func testRunnerLogger(
 	shout(ctx, l, os.Stdout, "test runner logs in: %s", runnerLogPath)
 	return l, teeOpt
 }
+
+// runOperation sequentially runs one operation matched by the passed-in filter.
+func runOperation(register func(registry.Registry), filter string, clusterName string) error {
+	//lint:ignore SA1019 deprecated
+	rand.Seed(roachtestflags.GlobalSeed)
+	r := makeTestRegistry()
+
+	register(&r)
+	ctx := context.Background()
+	stopper := stop.NewStopper()
+	defer stopper.Stop(ctx)
+	// NB: root logger with no path always tees to Stdout.
+	l, err := logger.RootLogger("", logger.NoTee)
+	if err != nil {
+		return err
+	}
+	// TODO(bilal): This is excessive for just getting the number of nodes in the
+	// cluster. We should expose a roachprod.Nodes method or so.
+	nodes, err := roachprod.PgURL(ctx, l, clusterName, roachtestflags.CertsDir, roachprod.PGURLOptions{})
+	if err != nil {
+		return errors.Wrap(err, "roachtest: run-operation: error when getting number of nodes")
+	}
+
+	cSpec := spec.ClusterSpec{NodeCount: len(nodes)}
+	c := &clusterImpl{
+		name:       clusterName,
+		spec:       cSpec,
+		l:          l,
+		expiration: cSpec.Expiration(),
+		destroyState: destroyState{
+			owned: false,
+		},
+		localCertsDir: roachtestflags.CertsDir,
+	}
+
+	specs, err := opsToRun(r, filter)
+	if err != nil {
+		return err
+	}
+	var opSpec *registry.OperationSpec
+	if len(specs) > 1 {
+		opSpec = &specs[rand.Intn(len(specs))]
+		l.Printf("more than one operation found for filter %s, randomly selected %s to run", filter, opSpec.Name)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	// Cancel this context if we get an interrupt.
+	CtrlC(ctx, l, cancel, nil /* registry */)
+	// Install goroutine leak checker and run it at the end of the entire operation
+	// run. This is good hygiene for operations, as operations can one day be
+	// called from roachtests as well.
+	defer leaktest.AfterTest(l)()
+
+	op := &operationImpl{
+		spec:         opSpec,
+		cockroach:    roachtestflags.CockroachBinaryPath,
+		l:            l,
+		cleanupState: make(map[string]string),
+	}
+	op.mu.cancel = cancel
+	var cleanup registry.OperationCleanup
+	{
+		ctx, cancel = context.WithTimeout(ctx, opSpec.Timeout)
```

Comment:
I personally don't see the value of this block. In fact, it makes it _more_ confusing, as it looks like `ctx` and `cancel` are scoped to the block, but in fact this statement is overwriting the outer variables.

In practice, this means we're also binding the cleanup function down below to this same context / timeout.

Reply:
Was meant to be a Function and with new ctx/cancel. Fixed.